### PR TITLE
feat: add native Grad-CAM XAI visualization and feature dominance mapping

### PR DESCRIPTION
**Pull request by FaresAminu**
### Summary
This PR introduces a native Explainable AI (XAI) feature dominance mapping utility directly into the Ultralytics ecosystem. It provides a standardized pipeline for generating targeted Grad-CAM heatmaps and quantitatively evaluating spatial feature attribution using Deletion and Insertion metrics.

### Motivation
As YOLO models are increasingly deployed in high-stakes scientific, diagnostic, and academic domains, simply providing a bounding box and a confidence score is often insufficient for peer review. Researchers require mathematical proof regarding *which specific spatial features* a model relied on to make a prediction. 

This utility allows users to extract the internal logic of the YOLO architecture, projecting high-dimensional bottleneck gradients and activations into human-readable 2D heatmaps, and generating Area Under the Curve (AUC) scores for faithfulness.

### Key Features and Changes
*   **Targeted Hook Management:** New `YOLO_XAI_Extractor` class for cleanly attaching and removing PyTorch forward and backward hooks to specific YOLO sequential layers.
*   **Targeted Backpropagation:** Mathematical functions to isolate and backpropagate the exact scalar confidence value from a single targeted anchor box, ensuring the explanation maps strictly to that specific detection.
*   **Grad-CAM & Metric Integration:** Optimized mathematical implementation of the Grad-CAM flow and native `validate_heatmap` functionality to automatically compute the Area Under Deletion Curve (AUDC) and Area Under Insertion Curve (AUIC).

### Checkboxes
- [x] I have read the CLA Document and I sign the CLA.
- [x] My contribution is licensed under the AGPL-3.0 License.
- [x] Code adheres to the Ruff Formatter standards (standardized 120 char limit).
- [x] New functions include Google-style docstrings.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds native Grad-CAM explainability utilities for extracting YOLO feature activations and gradients, generating heatmaps, and measuring visualization faithfulness.

### 📊 Key Changes
- Introduces `YOLO_XAI_Extractor` with forward and backward hooks for a configurable YOLO layer.
- Adds Grad-CAM heatmap generation with upsampling and normalization.
- Adds deletion and insertion confidence metrics for heatmap validation.
- Uses OpenCV, NumPy, PyTorch, and PyTorch functional APIs for visualization and evaluation.

### 🎯 Purpose & Impact
- Enables users to inspect which image regions influence YOLO predictions.
- Provides feature dominance mapping and quantitative heatmap evaluation for explainable AI workflows.
- May require follow-up integration, documentation, and tests to ensure compatibility across model architectures and inference scenarios.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @FaresAminu, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution has been received, and an engineer will assist with the review. Please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/utils/xai.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+
+class YOLO_XAI_Extractor:
+    """A utility class to extract feature maps and gradients from YOLO bottlenecks for Explainable AI (XAI) feature
+    dominance mapping (Grad-CAM).
+    """
+
+    def __init__(self, model, target_layer_index=22):
+        self.model = model.model
+        self.target_layer = self.model.model[target_layer_index]
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/xai.py]: ⚠️ **HIGH**: The default index only exists for the three-scale detection/segment-style graphs. Supported YOLO26 classification models have layers 0–10, and semantic segmentation models have layers 0–17, so constructing this extractor with its documented default raises `IndexError`. Select the penultimate layer dynamically (or require and validate a task-appropriate layer) rather than hard-coding the detector-specific index.

**Suggested change:**
```suggestion
        self.target_layer = self.model.model[target_layer_index if target_layer_index is not None else -2]
```

File: ultralytics/utils/xai.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    blurred_baseline = cv2.GaussianBlur(img_bgr, (51, 51), 0)
+
+    for pct in percentiles:
+        # 1. Deletion Masking
+        if pct == 100:
+            masked_img = np.zeros_like(img_bgr)
+        elif pct == 0:
+            masked_img = img_bgr.copy()
+        else:
+            thr = np.percentile(heatmap, pct)
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/xai.py]: ⚠️ **HIGH**: Deletion is ordered backwards: at `pct=5`, this uses the 5th-percentile threshold and masks nearly 95% of the highest-attribution pixels; at `pct=95`, it masks only about 5%. The deletion curve therefore measures restoration as the percentile increases, making the reported AUC inconsistent with the stated Deletion metric. Use the complementary percentile, as the insertion branch already does.

**Suggested change:**
```suggestion
            thr = np.percentile(heatmap, 100 - pct)
```

File: ultralytics/utils/xai.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            thr = np.percentile(heatmap, 100 - pct)
+            mask = (heatmap >= thr).astype(np.uint8)
+            inserted_img = (img_bgr * mask[..., np.newaxis] + blurred_baseline * (1 - mask[..., np.newaxis])).astype(
+                np.uint8
+            )
+
+        insertion_confs.append(get_confidence(inserted_img))
+
+    # Calculate Area Under the Curve (normalized to [0,1])
+    audc = np.trapezoid(deletion_confs, percentiles) / 100.0
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/xai.py]: ⚠️ **HIGH**: The project supports NumPy 1.23+, but `np.trapezoid` is unavailable before NumPy 2.0, so both AUC calculations raise `AttributeError` in supported environments. `utils/metrics.py` already handles this compatibility boundary with an `np.trapz` fallback; apply the same pattern here for both calculations.

File: ultralytics/utils/xai.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        # 1. Deletion Masking
+        if pct == 100:
+            masked_img = np.zeros_like(img_bgr)
+        elif pct == 0:
+            masked_img = img_bgr.copy()
+        else:
+            thr = np.percentile(heatmap, pct)
+            mask = (heatmap > thr).astype(np.uint8)
+            deletion_mask = 1 - mask
+            masked_img = (img_bgr * deletion_mask[..., np.newaxis]).astype(np.uint8)
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/xai.py:116]: 💡 **MEDIUM**: `generate_gradcam_heatmap()` defaults to a 640×640 map, while `img_bgr` is documented as the original image and can have any dimensions. For a normal non-640×640 source, this multiplication raises a broadcasting error before any metric is produced. The validation path needs to resize/project the CAM to the perturbed image dimensions, including the model letterbox transform, before creating masks.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The Grad-CAM computation is compact, but the current defaults and metric implementation have supported-model compatibility and faithfulness-scoring failures that need correction before this can be relied on.

💬 Posted 4 inline comments

**Comment by FaresAminu** [comment]
I have read the CLA Document and I sign the CLA

**Comment by FaresAminu** [comment]
I have read the CLA Document and I sign the CLA

**Comment by FaresAminu** [comment]
All automated review comments are resolved, the docstring has been added, and CI tests are passing (the Docs build failure appears unrelated). Could a maintainer please review when available? Thank you.

**Comment by glenn-jocher** [comment]
Thanks, CI is green and this is queued for maintainer review. Please add focused tests—current patch coverage is 0%—and user-facing documentation, and clarify how this API complements the existing `class_activation_map` utility.

**Comment by FaresAminu** [comment]
Thank you for the review. The focused tests have now been added, and the latest CI run passes across all required environments with 100% patch coverage.

Regarding the relationship with the existing class_activation_map utility: this contribution is intended to complement the existing functionality by providing a YOLO-specific Grad-CAM extraction workflow with configurable internal target-layer hooks, together with quantitative Deletion/Insertion faithfulness evaluation. I will also add/clarify the user-facing documentation to make this distinction explicit.

The PR is now green and ready for maintainer review. Thank you.

**Comment by glenn-jocher** [comment]
Thanks, the focused tests and API distinction look clear. Please push the user-facing documentation, including a minimal usage example and comparison with [`class_activation_map`](https://docs.ultralytics.com/reference/utils/plotting#ultralytics.utils.plotting.class_activation_map), then we can continue maintainer review.

**Comment by glenn-jocher** [comment]
Check this PR again, I think we already have this integrated via a recent @Y-T-G PR?

**Review by Y-T-G — COMMENTED** [review]
Hello @FaresAminu 

We have recently merged a heatmap visualization feature already in https://github.com/ultralytics/ultralytics/pull/25548 that uses LayerCAM. You can use `visualize=True` to see the heatmap on the image.

The PR description also explains why GradCAM is unsuitable for object detection models. The LayerCAM implementation is able to obtain better heatmap because it doesn't use only a single layer, but multiple layers to get the heatmap which works better with YOLO because we have multiple outputs from different scales.

Here's a comparison of the output of `visualize=True` vs. the output from this PR

| `visualize=True` | PR |
  | :---: | :---: |
  | bus_cam | image |


You can use the `visualize=True` feature instead. It works a lot better. It also supports filtering specific classes through `classes` argument.

We also [recommend starting with shorter and non-feature PRs](https://github.com/ultralytics/ultralytics/blob/main/CONTRIBUTING.md#first-time-contributors) as a new contributor. You can also check out our [contributing guide](https://docs.ultralytics.com/help/contributing) for more guidance.
